lib/ansiblelint/errors.py

```diff
@@ -1,37 +1,27 @@
 """Exceptions and error representations."""
 
 
-class Match(object):
-    """Rule violation detected during linting."""
+class Match(Exception):
+    """Rule violation detected during linting.
 
-    def __init__(self, linenumber, line, filename, rule, message=None) -> None:
+    It can be raised as Exception but also just added to the list of found
+    rules violations.
+    """
+
+    def __init__(self, message=None, linenumber=0, line=None, filename=None, rule=None) -> None:
         """Initialize a Match instance."""
+        super().__init__(message)
+
+        self.message = message or getattr(rule, 'shortdesc', "")
         self.linenumber = linenumber
         self.line = line
         self.filename = filename
         self.rule = rule
-        self.message = message or rule.shortdesc
 
     def __repr__(self):
         """Return a Match instance representation."""
         formatstr = u"[{0}] ({1}) matched {2}:{3} {4}"
-        return formatstr.format(self.rule.id, self.message,
-                                self.filename, self.linenumber, self.line)
-
-
-class MatchError(ValueError):
-    """Exception that would end-up as linter rule match."""
+        _id = getattr(self.rule, "id", "000")
```

Comment:
Rules have int ids, no need to fall-back to a string. Besides, you can easily use the number padding specifier:
```suggestion
        formatstr = "[{0:03d}] ({1}) matched {2}:{3} {4}"
        _id = getattr(self.rule, "id", 0)
```

Reply:
`rule.id` can be int, str or even missing, @albinvass wrote some custom rules for zuul starting with "ZUUL" prefix and I do not want to break them.

Reply:
@webknjaz yes, that's even part of the documentation for creating custom rules:
https://docs.ansible.com/ansible-lint/rules/rules.html#creating-custom-rules